src/haz3lcore/zipper/Projector.re

```diff
@@ -0,0 +1,315 @@
+      ({relatives: {siblings: (ls, rs), _}, _} as z: ZipperBase.t) => {
+    let (ls, content) = Segment.push_right((ls, z.selection.content));
+    z
+    |> ZipperBase.put_selection_content(content)
+    |> ZipperBase.put_siblings((ls, rs))
+    |> Option.some;
+  };
+
+  let skip_grow_right =
+      ({relatives: {siblings: (ls, rs), _}, _} as z: ZipperBase.t) => {
+    let (content, rs) = Segment.push_left((z.selection.content, rs));
+    z
+    |> ZipperBase.put_selection_content(content)
+    |> ZipperBase.put_siblings((ls, rs))
+    |> Option.some;
+  };
+
+  let grow =
+      ({relatives: {siblings, _}, projectors, _} as z: ZipperBase.t)
+      : option(ZipperBase.t) =>
+    switch (z.selection.focus, neighbor_is(projectors, siblings)) {
+    | (Left, (Some(_), _)) => skip_grow_left(z)
+    | (Right, (_, Some(_))) => skip_grow_right(z)
+    | _ => None
+    };
+
+  let skip_shrink_left =
+      ({relatives: {siblings: (ls, rs), _}, _} as z: ZipperBase.t) => {
+    let (ls, content) = Segment.push_left((ls, z.selection.content));
+    z
+    |> ZipperBase.put_selection_content(content)
+    |> ZipperBase.put_siblings((ls, rs))
+    |> Option.some;
+  };
+
+  let skip_shrink_right =
+      ({relatives: {siblings: (ls, rs), _}, _} as z: ZipperBase.t) => {
+    let (content, rs) = Segment.push_right((z.selection.content, rs));
+    z
+    |> ZipperBase.put_selection_content(content)
+    |> ZipperBase.put_siblings((ls, rs))
+    |> Option.some;
+  };
+
+  let selection_sides_is =
+      (projectors, s: Selection.t): (option(Id.t), option(Id.t)) => (
+    syntax_is(projectors, ListUtil.hd_opt(s.content)),
+    syntax_is(projectors, ListUtil.last_opt(s.content)),
+  );
+
+  let shrink =
+      ({selection, projectors, _} as z: ZipperBase.t): option(ZipperBase.t) =>
+    switch (selection.focus, selection_sides_is(projectors, selection)) {
+    | (Left, (Some(_), _)) => skip_shrink_left(z)
+    | (Right, (_, Some(_))) => skip_shrink_right(z)
+    | _ => None
+    };
+};
+
+/* See Select description above */
+module Move = {
+  let go = (d: Direction.t, z: ZipperBase.t): option(ZipperBase.t) =>
+    switch (d, neighbor_is(z.projectors, z.relatives.siblings)) {
+    | (Left, (Some(_), _)) =>
+      Some(
+        ZipperBase.put_siblings(Segment.push_right(z.relatives.siblings), z),
+      )
+    | (Right, (_, Some(_))) =>
+      Some(
+        ZipperBase.put_siblings(Segment.push_left(z.relatives.siblings), z),
+      )
+    | _ => None
+    };
+};
+
+module MapPiece = {
+  type updater = Piece.t => Piece.t;
+
+  let rec of_segment = (f: updater, seg: Segment.t): Segment.t => {
+    seg |> List.map(p => f(p)) |> List.map(of_piece(f));
+  }
+  and of_piece = (f: updater, piece: Piece.t): Piece.t => {
+    switch (piece) {
+    | Tile(t) => Tile(of_tile(f, t))
+    | Grout(_)
+    | Secondary(_) => piece
+    };
+  }
+  and of_tile = (f: updater, t: Tile.t): Tile.t => {
+    {...t, children: List.map(of_segment(f), t.children)};
+  };
```

Comment:
I'm still trying to understand exactly what this module is doing but could the updater just be on `tile -> tile` as opposed to `Piece.t -> Piece.t`?